### ISSUE TITLE
Consider multi-mesh API?

### ISSUE THREAD

**Issue by donmccurdy** [wwwtyro/geo-ambient-occlusion#3]
Had the chance to try this tool out today, and was quite happy with the result:

| before | after |
|---|---|
|  ![before](https://user-images.githubusercontent.com/1848368/46165957-d9a88780-c246-11e8-9a8c-53125f8e1305.png) | ![after](https://user-images.githubusercontent.com/1848368/46165966-dd3c0e80-c246-11e8-9e80-5e745b04fa8e.png) |

I'm thinking of using geo-ambient-occlusion as the starting point for a lightweight browser-based AO baking GUI, where for example a user could drag in a glTF model, bake AO, adjust settings, and export the result. To make that useful, I'd need to be able to support models containing several meshes and materials, and have each mesh occlude the others.

What do you think about adding a multi-mesh API:

```js
const geoao = require('geo-ambient-occlusion');

const aoSampler = geoao({resolution})
  .add(positions1, {transform1, cells1})
  .add(positions2, {transform2, cells2})
  .add(positions3, {transform3, cells3});

for (let i = 0; i < 256; i++) aoSampler.sample();

const [ao1, ao2, ao3] = aoSampler.report();
```

This can be accomplished without modifying the library (by packing and unpacking vertices into a single array) but I could imagine it being useful to others, and wanted to check before I implemented it in application code. If you don't have the time to do something like this, would you be interested in a PR? Or does it seem too complex for the scope of this project?

**Comment by wwwtyro** [comment]
That looks great! Both your result and your proposed API. If you don't give me a PR, I'll probably do it myself. :+1:

**Comment by donmccurdy** [comment]
Thanks, great to hear you're open to the idea! It's likely to be at least a few weeks before I could look into this seriously. I'll comment here before I start in case you decide to give it a shot. :)

**Comment by quacrobat** [comment]
Hello, has this been implemented in the end? also, how would it work with Three.js? 
thanks

**Comment by donmccurdy** [comment]
Sorry, I haven't found time to work on multi-mesh occlusion. I did bundle up my implementation for glTF files into this script:

https://github.com/donmccurdy/glTF-Transform/blob/main/packages/cli/src/transforms/ao.ts

```
npm install --global @gltf-transform/cli

gltf-transform ao input.glb output.glb
```

One mesh does not occlude another, as this issue describes, so that's a known limitation. glTF models created this way can be loaded in three.js normally, the (per-vertex) AO is written into UV coordinates referencing a 1x256 greyscale gradient, since three.js doesn't take AO as a vertex attribute.

**Comment by quacrobat** [comment]
Great, thanks for the reactivity and for sharing. 
I understand from what you write that multi-meshes are actually occluded but only by themselves, so it's not that multi-mesh files will simply not work with this tool. Is that correct?

**Comment by donmccurdy** [comment]
That's correct, yes – this tool can be used with files containing multiple meshes, but each mesh's occlusion is computed in isolation.
